daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/package.scala

```diff
@@ -23,4 +28,25 @@ package object transaction {
       }
       Right(b.result())
     }
+
+  sealed abstract class DiscriminatedSubtype[X] {
+    type T <: X
+    def apply(x: X): T
+    def subst[F[_]](fx: F[X]): F[T]
+  }
+
+  object DiscriminatedSubtype {
+    def apply[X]: DiscriminatedSubtype[X] = new DiscriminatedSubtype[X] {
+      override type T = X
+      override def apply(x: X): T = x
+      override def subst[F[_]](fx: F[X]): F[T] = fx
+    }
+  }
```

Comment:
Don't put `class`es, `trait`s, or `object`s in a `package object`. They can just go in the package.

Reply:
Indeed.